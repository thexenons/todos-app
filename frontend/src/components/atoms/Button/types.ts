import { ComponentPropsWithoutRef } from "react";

export enum ButtonVariants {
	text = "text",
	contained = "contained",
	outlined = "outlined",
}

export enum ButtonColorVariants {
	primary = "primary",
	secondary = "secondary",
	neutral = "neutral",
}

export interface ButtonProps extends ComponentPropsWithoutRef<"button"> {
	variant?: ButtonVariants;
	colorVariant?: ButtonColorVariants;
	isRounded?: boolean;
}
