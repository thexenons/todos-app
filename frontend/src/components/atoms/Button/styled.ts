import { rgba } from "polished";
import styled, { css } from "styled-components";

import { ButtonColorVariants, ButtonVariants } from "./types";

const buttonVariants: Record<
	keyof typeof ButtonVariants,
	ReturnType<typeof css>
> = {
	text: css`
		color: var(--button-color-main);
		background: transparent;
		border-color: transparent;

		&:hover {
			background: var(--button-color-light-opacified);
		}

		&:disabled {
			background: transparent;
		}
	`,
	contained: css`
		color: var(--button-color-main-text);
		background: var(--button-color-main);
		border-color: var(--button-color-main);

		&:hover {
			color: var(--button-color-dark-text);
			background: var(--button-color-dark);
			border-color: var(--button-color-dark);
		}

		&:disabled {
			color: var(--button-color-main-text);
			background: var(--button-color-main);
			border-color: var(--button-color-main);
		}
	`,
	outlined: css`
		background: transparent;
		color: var(--button-color-main);
		border-color: var(--button-color-light);

		&:hover {
			background: var(--button-color-light-opacified);
			border-color: var(--button-color-main);
		}

		&:disabled {
			background: transparent;
			border-color: var(--button-color-light);
		}
	`,
};

const buttonWrapperVariants: Record<
	keyof typeof ButtonColorVariants,
	ReturnType<typeof css>
> = {
	[ButtonColorVariants.primary]: css`
		--button-color-light-opacified: ${({ theme }) =>
			rgba(theme.colors.primary.light, 0.1)};
		--button-color-light: ${({ theme }) => theme.colors.primary.light};
		--button-color-light-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.primary.light)};
		--button-color-main: ${({ theme }) => theme.colors.primary.main};
		--button-color-main-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.primary.main)};
		--button-color-dark: ${({ theme }) => theme.colors.primary.dark};
		--button-color-dark-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.primary.dark)};
	`,
	[ButtonColorVariants.secondary]: css`
		--button-color-light-opacified: ${({ theme }) =>
			rgba(theme.colors.secondary.light, 0.1)};
		--button-color-light: ${({ theme }) => theme.colors.secondary.light};
		--button-color-light-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.secondary.light)};
		--button-color-main: ${({ theme }) => theme.colors.secondary.main};
		--button-color-main-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.secondary.main)};
		--button-color-dark: ${({ theme }) => theme.colors.secondary.dark};
		--button-color-dark-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.secondary.dark)};
	`,
	[ButtonColorVariants.neutral]: css`
		--button-color-light-opacified: ${({ theme }) =>
			rgba(theme.colors.grey.light, 0.1)};
		--button-color-light: ${({ theme }) => theme.colors.grey.light};
		--button-color-light-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.grey.light)};
		--button-color-main: ${({ theme }) => theme.colors.grey.main};
		--button-color-main-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.grey.main)};
		--button-color-dark: ${({ theme }) => theme.colors.grey.dark};
		--button-color-dark-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.grey.dark)};
	`,
	[ButtonColorVariants.error]: css`
		--button-color-light-opacified: ${({ theme }) =>
			rgba(theme.colors.error.light, 0.1)};
		--button-color-light: ${({ theme }) => theme.colors.error.light};
		--button-color-light-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.error.light)};
		--button-color-main: ${({ theme }) => theme.colors.error.main};
		--button-color-main-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.error.main)};
		--button-color-dark: ${({ theme }) => theme.colors.error.dark};
		--button-color-dark-text: ${({ theme }) =>
			theme.colors.getContrastText(theme.colors.error.dark)};
	`,
};

export const ButtonWrapper = styled.button<{
	$variant?: keyof typeof ButtonVariants;
	$colorVariant?: keyof typeof ButtonColorVariants;
	$isRounded?: boolean;
}>`
	--button-size: 36px;
	appearance: none;
	border: 1px solid transparent;
	min-height: var(--button-size);
	min-width: var(--button-size);
	${({ theme }) =>
		theme.mixins.flex({
			inline: true,
			align: "center",
			justify: "center",
			gap: theme.spacing(),
		})}
	${({ theme }) => theme.mixins.padding({ x: 1 })}
	${({ theme }) => theme.typography.getTypographyStyle(theme.typography.button)}
	transition: background 0.3s, color 0.3s, border-color 0.3s;
	border-radius: ${({ $isRounded, theme }) =>
		$isRounded ? "calc(var(--button-size) / 2)" : theme.spacing(0.5)};
	cursor: pointer;

	${({ $colorVariant = ButtonColorVariants.neutral }) =>
		buttonWrapperVariants[$colorVariant]}
	${({ $variant = ButtonVariants.contained }) => buttonVariants[$variant]};

	&:disabled {
		cursor: not-allowed;
		opacity: 0.7;
	}
`;

export const ButtonIconsWrapper = styled.span`
	${({ theme }) =>
		theme.mixins.flex({
			inline: true,
			justify: "center",
			align: "center",
			gap: theme.spacing(),
		})}
`;
