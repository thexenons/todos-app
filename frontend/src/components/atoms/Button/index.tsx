import { FC } from "react";

import * as S from "./styled";
import { ButtonColorVariants, ButtonProps, ButtonVariants } from "./types";

const Button: FC<ButtonProps> = ({
	children,
	variant = ButtonVariants.contained,
	colorVariant = ButtonColorVariants.neutral,
	isRounded = false,
	leftIcon,
	rightIcon,
	...rest
}) => (
	<S.ButtonWrapper
		$variant={variant}
		$colorVariant={colorVariant}
		$isRounded={isRounded}
		{...rest}
	>
		{leftIcon && <S.ButtonIconsWrapper>{leftIcon}</S.ButtonIconsWrapper>}
		{children}
		{rightIcon && <S.ButtonIconsWrapper>{rightIcon}</S.ButtonIconsWrapper>}
	</S.ButtonWrapper>
);

export default Button;
