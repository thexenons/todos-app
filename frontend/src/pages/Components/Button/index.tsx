import { Children, FC } from "react";
import { MdAlarmAdd, MdDelete, MdModeEdit } from "react-icons/md";

import Button from "../../../components/atoms/Button";
import {
	ButtonColorVariants,
	ButtonProps,
	ButtonVariants,
} from "../../../components/atoms/Button/types";
import * as S from "./styled";

const roundedVariants = [
	{
		title: "Default",
	},
	{
		title: "Rounded",
		isRounded: true,
	},
	{
		title: <MdAlarmAdd />,
	},
	{
		title: <MdAlarmAdd />,
		isRounded: true,
	},
];

const propsVariants = [
	{
		title: "leftIcon",
		leftIcon: Children.toArray([<MdModeEdit />, <MdDelete />]),
	},
	{
		title: "rightIcon",
		rightIcon: Children.toArray([<MdModeEdit />, <MdDelete />]),
	},
	{
		title: "both icons",
		leftIcon: <MdModeEdit />,
		rightIcon: <MdModeEdit />,
	},
];

const buttons = roundedVariants.reduce<ButtonProps[]>(
	(prev, { title, ...restRoundedVariant }) => {
		prev.push({
			children: title,
			...restRoundedVariant,
		});

		if (typeof title === "string") {
			propsVariants.forEach((propsVariant) => {
				prev.push({
					children: `${title} ${propsVariant.title}`,
					...restRoundedVariant,
					...propsVariant,
				});
			});
		}

		return prev;
	},
	[]
);

const buttonsLists = buttons.reduce<ButtonProps[][]>((prev, curr, index) => {
	if (!prev[index]) prev[index] = [];

	prev[index].push(curr);
	prev[index].push({ ...curr, disabled: true });

	return prev;
}, []);

const ButtonComponentsPage: FC = () => (
	<S.ButtonListWrapper>
		<h1>Buttons</h1>
		{Children.toArray(
			Object.keys(ButtonVariants).map((buttonVariant) => (
				<S.ButtonListWrapper>
					<h2>{buttonVariant}</h2>
					{Children.toArray(
						Object.keys(ButtonColorVariants).map((buttonColorVariant) => (
							<S.ButtonListWrapper>
								<h3>{buttonColorVariant}</h3>
								{Children.toArray(
									buttonsLists.map((buttonsList) => (
										<S.ButtonsWrapper>
											{Children.toArray(
												buttonsList.map(({ children, ...restButton }) => (
													<Button
														variant={buttonVariant as ButtonVariants}
														colorVariant={
															buttonColorVariant as ButtonColorVariants
														}
														{...restButton}
													>
														{children}
													</Button>
												))
											)}
										</S.ButtonsWrapper>
									))
								)}
							</S.ButtonListWrapper>
						))
					)}
				</S.ButtonListWrapper>
			))
		)}
	</S.ButtonListWrapper>
);

export default ButtonComponentsPage;
