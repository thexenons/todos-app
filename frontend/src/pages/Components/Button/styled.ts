import styled from "styled-components";

export const ButtonListWrapper = styled.div`
	${({ theme }) =>
		theme.mixins.flex({ direction: "column", gap: theme.spacing(2) })}
`;

export const ButtonsWrapper = styled.div`
	${({ theme }) => theme.mixins.flex({ gap: theme.spacing(2) })}
`;
