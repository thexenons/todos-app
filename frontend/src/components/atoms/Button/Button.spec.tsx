import { render } from "@testing-library/react";

import BaseTestComponent from "../../../tests/BaseTestComponent";
import Button from ".";

describe("Button", () => {
	it("Render", () => {
		const { container } = render(
			<BaseTestComponent>
				<Button />
			</BaseTestComponent>
		);
		expect(container).toMatchSnapshot();
	});
});
