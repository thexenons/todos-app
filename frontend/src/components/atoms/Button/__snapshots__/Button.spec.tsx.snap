// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > Render 1`] = `
<div>
  <button
    class="styled__ButtonWrapper-sc-1fzdj0n-0 fgaVpy"
  />
</div>
`;
